sbin/route/route.8
.\" Copyright (c) 1983 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" %sccs.include.redist.man%
.\"
.\"	@(#)route.8	6.9 (Berkeley) 7/26/90
.\"
.TH ROUTE 8 ""
.UC 5
.SH NAME
route \- manually manipulate the routing tables (Interim)
.SH SYNOPSIS
.B route
[
.B \-n
] [
.B \-q
] [
.B \-v
]
.I command
[ [
.I modifiers
]
.I args
]
.SH DESCRIPTION
.I Route
is a program used to manually manipulate the network
routing tables.  It normally is not needed, as the
system routing table management daemon,
.IR routed (8C),
should tend to this task.
.PP
.I Route
accepts five commands:
.IR add ,
to add a route,
.IR flush ,
to remove all routes,
.IR delete ,
to delete a specific route,
.IR change ,
to changes aspects of a route (such as its gateway),
and
.IR monitor ,
to report any changes to the routing information base,
routing lookup misses, or suspected network partionings.
.PP
The monitor command has the syntax
.ti +0.25i
.B route 
[
.B -n 
]
.B monitor
.PP
The flush command has the syntax
.ti +0.25i
.B route
[
.B -n
]
[
.B -n
]
.B flush
[
.I family
]
.PP
where the address family may be specified by any of the
.BR -osi ,
.BR -xns ,
or
.B -inet
keywords.
.PP
The other commands have the following syntax:
.PP
.ti +0.25i
.B route 
[
.I -n 
]
.I command
[
.B -net
|
.B -host
]
.I destination gateway
.PP
where
.I destination
is the destination host or network,
.I gateway
is the next-hop gateway to which packets should be addressed.
Routes to a particular host are distinguished from those to
a network by interpreting the Internet address associated with
.IR destination .
The optional keywords
.B -net
and
.B -host
force the destination to be interpreted as a network or a host, respectively.
Otherwise, if the 
.I destination
has a ``local address part'' of INADDR_ANY,
or if the
.I destination
is the symbolic name of a network, then the route is
assumed to be to a network; otherwise, it is presumed to be a
route to a host.
.PP
For example,
128.32 is interpreted as
.I -host
128.0.0.32,
128.32.130 is interpreted as
.I -host
128.32.0.130;
.I -net 128.32
is interpreted as
128.32.0.0,
and 
.I -net 128.32.130
is interpreted as
128.32.130.0.
.PP
If the route is via an interface rather than
via a gateway, the 
.I -interface
modifier should be specified;
the gateway given is the address of this host on the common network,
indicating the interface to be used for transmission.
.PP
The optional modifiers
.I -xns,
and
.I -osi
specify that all subsequent addresses are in the XNS or OSI address families,
and the names must be numeric specifications rather than
symbolic names.
.PP
The optional
.I -netmask
qualifier is intended
to acheieve the effect of an OSI ESIS redirect with the netmask option.
One specifies an additional ensuing address parameter
(to be interpreted as a network mask).
One can override the implicit network mask generated in the inet case
by making sure this option follows the destination parameter.
.PP
The optional modifiers
.IR -rtt ,
.IR -rttvar ,
.IR -sendpipe ,
.IR -recvpipe ,
.IR -mtu ,
.IR -hopcount ,
.IR -expire ,
and
.I -ssthresh 
provide initial values to metrics maintained in the routing entry.
These may be individually locked by preceding each such modifier to
be locked by
the
.I -lock
meta-modifier, or one can 
specify that all ensuing metrics may be locked by the
.I -lockrest
meta-modifier.
.PP
All symbolic names specified for a
.I destination 
or 
.I gateway
are looked up first as a host name using
.IR gethostbyname (3N).
If this lookup fails,
.IR getnetbyname (3N)
is then used to interpret the name as that of a network.
.PP
.I Route
uses a routing socket and the new message types
RTM_ADD, RTM_DELETE, and RTM_CHANGE.
As such, only the super-user may modify
the routing tables.
.PP
If the 
.B flush
command is specified, 
.I route
will ``flush'' the routing tables of all gateway entries.
One can choose to flush only those routes whose destinations
are of a given address family, by specifying an optional keyword
describing which address family.
.PP
The
.B \-n
option prevents attempts to print host and network names symbolically
when reporting actions.
The
.B \-v
option causes additional details to be printed.
The
.B \-q
option supresses all output.
.SH DIAGNOSTICS
``\fBadd [ host | network ] %s: gateway %s flags %x\fP''
.br
The specified route is being added to the tables.  The
values printed are from the routing table entry supplied
in the 
.I ioctl
call.
If the gateway address used was not the primary address of the gateway
(the first one returned by
.IR gethostbyname ),
the gateway address is printed numerically as well as symbolically.
.PP
``\fBdelete [ host | network ] %s: gateway %s flags %x\fP''
.br
As above, but when deleting an entry.
.PP
``\fB%s %s done\fP''
.br
When the 
.B \-f
flag is specified, each routing table entry deleted
is indicated with a message of this form.
.PP
``\fBNetwork is unreachable\fP''
.br
An attempt to add a route failed because the gateway listed was not
on a directly-connected network.
The next-hop gateway must be given.
.PP
``\fBnot in table\fP''
.br
A delete operation was attempted for an entry which
wasn't present in the tables.
.PP
``\fBrouting table overflow\fP''
.br
An add operation was attempted, but the system was
low on resources and was unable to allocate memory
to create the new entry.
.SH "SEE ALSO"
netintro(4), route(4), esis(4), routed(8), XNSrouted(8)


sbin/route/route.c
/*
 * Copyright (c) 1983 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1983 Regents of the University of California.\n\
 All rights reserved.\n";
#endif not lint

#ifndef lint
static char sccsid[] = "@(#)route.c	5.4 (Berkeley) 2/18/86";
#endif not lint

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <sys/mbuf.h>

#include <net/route.h>
#include <netinet/in.h>

#include <stdio.h>
#include <errno.h>
#include <ctype.h>
#include <netdb.h>

struct	rtentry route;
int	s;
int	forcehost, forcenet;
struct	sockaddr_in sin = { AF_INET };
struct	in_addr inet_makeaddr();
char	*malloc();

main(argc, argv)
	int argc;
	char *argv[];
{

	if (argc < 2)
		printf("usage: route [ -f ] [ [ -h ] [ -n ] cmd args ]\n"),
		exit(1);
	s = socket(AF_INET, SOCK_RAW, 0);
	if (s < 0) {
		perror("route: socket");
		exit(1);
	}
	argc--, argv++;
	for (; argc >  0 && argv[0][0] == '-'; argc--, argv++) {
		for (argv[0]++; *argv[0]; argv[0]++)
			switch (*argv[0]) {
			case 'f':
				flushroutes();
				break;
			case 'h':
				forcehost++;
				break;
			case 'n':
				forcenet++;
				break;
			}
	}
	if (forcehost && forcenet) {
		fprintf(stderr, "-n and -h are incompatible\n");
		exit(1);
	}
	if (argc > 0) {
		if (strcmp(*argv, "add") == 0)
			newroute(argc, argv);
		else if (strcmp(*argv, "delete") == 0)
			newroute(argc, argv);
		else if (strcmp(*argv, "change") == 0)
			changeroute(argc-1, argv+1);
		else
			printf("%s: huh?\n", *argv);
	}
}

/*
 * Purge all entries in the routing tables not
 * associated with network interfaces.
 */
#include <nlist.h>

struct nlist nl[] = {
#define	N_RTHOST	0
	{ "_rthost" },
#define	N_RTNET		1
	{ "_rtnet" },
#define N_RTHASHSIZE	2
	{ "_rthashsize" },
	"",
};

flushroutes()
{
	struct mbuf mb;
	register struct rtentry *rt;
	register struct mbuf *m;
	struct mbuf **routehash;
	int rthashsize, i, doinghost = 1, kmem;
	char *routename();

	nlist("/vmunix", nl);
	if (nl[N_RTHOST].n_value == 0) {
		printf("route: \"rthost\", symbol not in namelist\n");
		exit(1);
	}
	if (nl[N_RTNET].n_value == 0) {
		printf("route: \"rtnet\", symbol not in namelist\n");
		exit(1);
	}
	if (nl[N_RTHASHSIZE].n_value == 0) {
		printf("route: \"rthashsize\", symbol not in namelist\n");
		exit(1);
	}
	kmem = open("/dev/kmem", 0);
	if (kmem < 0) {
		perror("route: /dev/kmem");
		exit(1);
	}
	lseek(kmem, nl[N_RTHASHSIZE].n_value, 0);
	read(kmem, &rthashsize, sizeof (rthashsize));
	routehash = (struct mbuf **)malloc(rthashsize*sizeof (struct mbuf *));

	lseek(kmem, nl[N_RTHOST].n_value, 0);
	read(kmem, routehash, rthashsize*sizeof (struct mbuf *));
	printf("Flushing routing tables:\n");
again:
	for (i = 0; i < rthashsize; i++) {
		if (routehash[i] == 0)
			continue;
		m = routehash[i];
		while (m) {
			lseek(kmem, m, 0);
			read(kmem, &mb, sizeof (mb));
			rt = mtod(&mb, struct rtentry *);
			if (rt->rt_flags & RTF_GATEWAY) {
				struct sockaddr_in *sin;

				sin = (struct sockaddr_in *)&rt->rt_dst;
				printf("%-15.15s ", routename(sin->sin_addr));
				sin = (struct sockaddr_in *)&rt->rt_gateway;
				printf("%-15.15s ", routename(sin->sin_addr));
				if (ioctl(s, SIOCDELRT, (caddr_t)rt) < 0)
					error("delete");
				else
					printf("done\n");
			}
			m = mb.m_next;
		}
	}
	if (doinghost) {
		lseek(kmem, nl[N_RTNET].n_value, 0);
		read(kmem, routehash, rthashsize*sizeof (struct mbuf *));
		doinghost = 0;
		goto again;
	}
	close(kmem);
	free(routehash);
}

char *
routename(in)
	struct in_addr in;
{
	char *cp = 0;
	static char line[50];
	struct hostent *hp;
	struct netent *np;
	int lna, net;
	static int bynumber;

	net = inet_netof(in);
	lna = inet_lnaof(in);
	if (in.s_addr == INADDR_ANY)
		cp = "default";
	if (cp == 0 && (lna == INADDR_ANY || forcenet)) {
		np = getnetbyaddr(net, AF_INET);
		if (np)
			cp = np->n_name;
	}
	if (cp == 0 && bynumber == 0) {
		hp = gethostbyaddr(&in, sizeof (struct in_addr), AF_INET);
		if (hp)
			cp = hp->h_name;
		else if (h_errno == TRY_AGAIN)
			bynumber = 1;
	}
	if (cp)
		strcpy(line, cp);
	else {
		u_char *ucp = (u_char *)&in;
		if (lna == INADDR_ANY)
			sprintf(line, "%u.%u.%u", ucp[0], ucp[1], ucp[2]);
		else
			sprintf(line, "%u.%u.%u.%u", ucp[0], ucp[1],
				ucp[2], ucp[3]);
	}
	return (line);
}

newroute(argc, argv)
	int argc;
	char *argv[];
{
	struct sockaddr_in *sin;
	char *cmd, *dest, *gateway;
	int ishost, metric = 0, ret, attempts, oerrno;
	struct hostent *hp;
	extern int errno;

	cmd = argv[0];
	if (*cmd == 'a') {
		if (argc != 4) {
			printf("usage: %s destination gateway metric\n", cmd);
			printf("(metric of 0 if gateway is this host)\n");
			return;
		}
		metric = atoi(argv[3]);
	} else {
		if (argc < 3) {
			printf("usage: %s destination gateway\n", cmd);
			return;
		}
	}
	sin = (struct sockaddr_in *)&route.rt_dst;
	ishost = getaddr(argv[1], &route.rt_dst, &hp, &dest);
	if (forcehost)
		ishost = 1;
	if (forcenet)
		ishost = 0;
	sin = (struct sockaddr_in *)&route.rt_gateway;
	(void) getaddr(argv[2], &route.rt_gateway, &hp, &gateway);
	route.rt_flags = RTF_UP;
	if (ishost)
		route.rt_flags |= RTF_HOST;
	if (metric > 0)
		route.rt_flags |= RTF_GATEWAY;
	for (attempts = 1; ; attempts++) {
		errno = 0;
		if ((ret = ioctl(s, *cmd == 'a' ? SIOCADDRT : SIOCDELRT,
		     (caddr_t)&route)) == 0)
			break;
		if (errno != ENETUNREACH && errno != ESRCH)
			break;
		if (hp && hp->h_addr_list[1]) {
			hp->h_addr_list++;
			bcopy(hp->h_addr_list[0], (caddr_t)&sin->sin_addr,
			    hp->h_length);
		} else
			break;
	}
	oerrno = errno;
	printf("%s %s %s: gateway %s", cmd, ishost? "host" : "net",
		dest, gateway);
	if (attempts > 1 && ret == 0)
	    printf(" (%s)",
		inet_ntoa(((struct sockaddr_in *)&route.rt_gateway)->sin_addr));
	if (ret == 0)
		printf("\n");
	else {
		printf(": ");
		fflush(stdout);
		errno = oerrno;
		error(0);
	}
}

changeroute(argc, argv)
	int argc;
	char *argv[];
{
	printf("not supported\n");
}

error(cmd)
	char *cmd;
{

	if (errno == ESRCH)
		fprintf(stderr, "not in table\n");
	else if (errno == EBUSY)
		fprintf(stderr, "entry in use\n");
	else if (errno == ENOBUFS)
		fprintf(stderr, "routing table overflow\n");
	else
		perror(cmd);
}

char *
savestr(s)
	char *s;
{
	char *sav;

	sav = malloc(strlen(s) + 1);
	if (sav == NULL) {
		fprintf("route: out of memory\n");
		exit(1);
	}
	strcpy(sav, s);
	return (sav);
}

/*
 * Interpret an argument as a network address of some kind,
 * returning 1 if a host address, 0 if a network address.
 */
getaddr(s, sin, hpp, name)
	char *s;
	struct sockaddr_in *sin;
	struct hostent **hpp;
	char **name;
{
	struct hostent *hp;
	struct netent *np;
	u_long val;

	*hpp = 0;
	if (strcmp(s, "default") == 0) {
		sin->sin_family = AF_INET;
		sin->sin_addr = inet_makeaddr(0, INADDR_ANY);
		*name = "default";
		return(0);
	}
	sin->sin_family = AF_INET;
	val = inet_addr(s);
	if (val != -1) {
		sin->sin_addr.s_addr = val;
		*name = s;
		return(inet_lnaof(sin->sin_addr) != INADDR_ANY);
	}
	val = inet_network(s);
	if (val != -1) {
		sin->sin_addr = inet_makeaddr(val, INADDR_ANY);
		*name = s;
		return(0);
	}
	np = getnetbyname(s);
	if (np) {
		sin->sin_family = np->n_addrtype;
		sin->sin_addr = inet_makeaddr(np->n_net, INADDR_ANY);
		*name = savestr(np->n_name);
		return(0);
	}
	hp = gethostbyname(s);
	if (hp) {
		*hpp = hp;
		sin->sin_family = hp->h_addrtype;
		bcopy(hp->h_addr, &sin->sin_addr, hp->h_length);
		*name = savestr(hp->h_name);
		return(1);
	}
	fprintf(stderr, "%s: bad value\n", s);
	exit(1);
}
